edb/edgeql/compiler/stmt.py

```diff
@@ -228,86 +227,58 @@ def simple_stmt_eq(lhs: irast.Base, rhs: irast.Base,
         return False
 
 
-def handle_conditional_insert(
-        expr: qlast.InsertQuery,
-        rhs: irast.InsertStmt,
-        lhs_set: Union[irast.Expr, irast.Set],
+def compile_insert_on_conflict_constraint(
+        stmt: irast.InsertStmt,
+        constraint_spec: qlast.Expr,
         *, ctx: context.ContextLevel,
 ) -> irast.ConstraintRef:
-    def error(s: str) -> NoReturn:
+    ctx.partial_path_prefix = stmt.subject
+
+    # We compile the name here so we can analyze it, but we don't do
+    # anything else with it.
+    cspec_res = setgen.ensure_set(dispatch.compile(
+        constraint_spec, ctx=ctx), ctx=ctx)
+
+    if not cspec_res.rptr:
```

Comment:
This isn't enough, you should also check that: 1) `rptr.source` is the same as `InsertStmt` subject and that the cardinality of the property is `single`.

Reply:
And then consider raising `InvalidTypeError`.

Reply:
It doesn't seem like we raise InvalidTypeError much at all elsewhere? Am I missing something?

Reply:
I don't think we're too diligent with narrowing down error types, we need to improve with that. Most of the time we raise `QueryError` simply because it's a habit (and often a copy/paste). That said, our errors are hierarchical, and InvalidTypeError is a subclass of QueryError, and it seems that in this very case it's an appropriate one.

Reply:
BTW, `api/errors.txt` is not set in stone, so if you see that we're misusing QueryError in some common cases feel free to propose adding new types there.